# DeepMod2
DeepMod2 is a computational tool for detecting DNA methylation and modifications from Oxford Nanopore reads. It uses a BiLSTM model to predict per-read and per-site 5mC methylations for CpG sites. DeepMod2 can call methylation from FAST5 files basecalled with Guppy or resquiggled with Tombo. DeepMod2 now supports 5mC methylation calling from R10.4 flowcells.

DeepMod2 is distributed under the [MIT License by Wang Genomics Lab](https://wglab.mit-license.org/).


## Installation
Please refer to [Installation](https://github.com/WGLab/DeepMod2/blob/main/docs/Install.md) for how to install DeepMod2.

## Usage

Please refer to [Usage](https://github.com/WGLab/DeepMod2/blob/main/docs/Usage.md) for how to use DeepMod2.

## Examples

Please refer to [Example](https://github.com/WGLab/DeepMod2/blob/main/docs/Example.md) detailed instructions on how to run DeepMod2 on both Guppy and Tombo data.
